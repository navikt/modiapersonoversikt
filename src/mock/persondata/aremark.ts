import {
    AdresseBeskyttelse,
    EgenAnsatt,
    ForelderBarnRelasjon,
    ForelderBarnRelasjonRolle,
    FullmaktsRolle,
    Kjonn,
    LocalDate,
    LocalDateTime,
    Person,
    PersonStatus,
    SivilstandType,
    Skifteform
} from '../../app/personside/visittkort-v2/PersondataDomain';
import { harDiskresjonskode } from '../../app/personside/visittkort-v2/visittkort-utils';

export function hentBarnAremark(): ForelderBarnRelasjon[] {
    return [
        {
            ident: '12345678910',
            rolle: ForelderBarnRelasjonRolle.BARN,
            navn: [
                {
                    fornavn: 'BARN1',
                    mellomnavn: null,
                    etternavn: 'BARNESEN'
                }
            ],
            fodselsdato: ['2008-03-15' as LocalDate],
            alder: 13,
            kjonn: [
                {
                    kode: Kjonn.K,
                    beskrivelse: 'Kvinne'
                }
            ],
            adressebeskyttelse: [
                {
                    kode: AdresseBeskyttelse.UGRADERT,
                    beskrivelse: 'UGRADERT'
                }
            ],
            harSammeAdresse: false,
            dodsdato: []
        },
        {
            ident: '12345678911',
            rolle: ForelderBarnRelasjonRolle.BARN,
            navn: [
                {
                    fornavn: 'BARN1',
                    mellomnavn: null,
                    etternavn: 'BARNESEN'
                }
            ],
            fodselsdato: ['2008-03-15' as LocalDate],
            alder: null,
            kjonn: [
                {
                    kode: Kjonn.K,
                    beskrivelse: 'Kvinne'
                }
            ],
            adressebeskyttelse: [],
            harSammeAdresse: false,
            dodsdato: []
        }
    ];
}

export const aremark: Person = {
    fnr: '10108000398',
    navn: [
        {
            fornavn: 'AREMARK',
            mellomnavn: null,
            etternavn: 'TESTFAMILIEN'
        }
    ],
    kjonn: [
        {
            kode: Kjonn.M,
            beskrivelse: 'Mann'
        }
    ],
    fodselsdato: ['1980-10-10' as LocalDate],
    geografiskTilknytning: '0118',
    alder: 41,
    dodsdato: [],
    bostedAdresse: [
        {
            linje1: 'Islandsgate 49',
            linje2: '7693 Hovedøya',
            linje3: 'Norge',
            sistEndret: {
                ident: 'Folkeregisteret',
                tidspunkt: '2010-02-20T10:15:30' as LocalDateTime,
                system: 'FREG'
            },
            gyldighetsPeriode: null
        }
    ],
    kontaktAdresse: [
        {
            linje1: 'Aremarkveien 1',
            linje2: '0320 Aremark',
            linje3: null,
            sistEndret: {
                ident: 'Folkeregisteret',
                tidspunkt: '2020-01-01T10:15:30' as LocalDateTime,
                system: 'FREG'
            },
            gyldighetsPeriode: null
        }
    ],
    oppholdsAdresse: [
        {
            linje1: 'Oppholdsadresse 1',
            linje2: '0000 AREMARK',
            linje3: null,
            sistEndret: null,
            gyldighetsPeriode: {
                gyldigFraOgMed: '2021-02-01' as LocalDate,
                gyldigTilOgMed: '2022-05-01' as LocalDate
            }
        }
    ],
    navEnhet: {
        id: '0219',
        navn: 'NAV Bærum',
        publikumsmottak: [
            {
                besoksadresse: {
                    linje1: 'Adressevei 1',
                    linje2: '0000 AREMARK',
                    linje3: null,
                    sistEndret: null,
                    gyldighetsPeriode: null
                },
                apningstider: [
                    {
                        ukedag: 'Mandag',
                        apningstid: '09.00 - 15.00'
                    },
                    {
                        ukedag: 'Tirsdag',
                        apningstid: '09.00 - 15.00'
                    },
                    {
                        ukedag: 'Onsdag',
                        apningstid: '10.00 - 16.00'
                    }
                ]
            }
        ]
    },
    statsborgerskap: [
        {
            land: {
                kode: 'NOR',
                beskrivelse: 'NORGE'
            },
            gyldighetsPeriode: {
                gyldigFraOgMed: '2021-02-01' as LocalDate,
                gyldigTilOgMed: '2022-05-01' as LocalDate
            }
        }
    ],
    adressebeskyttelse: [
        {
            kode: AdresseBeskyttelse.KODE6,
            beskrivelse: 'Kode 6'
        }
    ],
    sikkerhetstiltak: [
        {
            type: 'TFUS',
            beskrivelse: 'Telefonisk utestengelse',
            gyldighetsPeriode: {
                gyldigFraOgMed: '2021-02-01' as LocalDate,
                gyldigTilOgMed: '2022-05-01' as LocalDate
            }
        }
    ],
    erEgenAnsatt: EgenAnsatt.JA,
    personstatus: [
        {
            kode: PersonStatus.BOSATT,
            beskrivelse: 'Bosatt i Norge'
        }
    ],
    sivilstand: [
        {
            type: {
                kode: SivilstandType.GIFT,
                beskrivelse: 'Gift'
            },
            gyldigFraOgMed: '2005-12-12' as LocalDate,
            sivilstandRelasjon: {
                fnr: '12345555555',
                navn: [
                    {
                        fornavn: 'BAREMARK',
                        mellomnavn: null,
                        etternavn: 'TESTFAMILIEN'
                    }
                ],
                alder: 40,
                adressebeskyttelse: [
                    {
                        kode: AdresseBeskyttelse.UGRADERT,
                        beskrivelse: 'UGRADERT'
                    }
                ],
                harSammeAdresse: true
            }
        }
    ],
    foreldreansvar: hentBarnAremark().map((barn) => ({
        ansvar: 'felles',
        ansvarlig: null,
        ansvarsubject: {
            navn: harDiskresjonskode(barn.adressebeskyttelse) ? null : barn.navn.firstOrNull(),
            ident: barn.ident
        }
    })),
    deltBosted: [
        {
            gyldighetsPeriode: {
                gyldigFraOgMed: '2021-02-01' as LocalDate,
                gyldigTilOgMed: '2022-05-01' as LocalDate
            },
            adresse: {
                linje1: `Adresseveien 1`,
                linje2: '0000 Aremark',
                linje3: null,
                sistEndret: null,
                gyldighetsPeriode: null
            }
        }
    ],
    dodsbo: [
        {
            adressat: {
                advokatSomAdressat: {
                    kontaktperson: {
                        fornavn: 'Advokat',
                        mellomnavn: null,
                        etternavn: 'Navnesen'
                    },
                    organisasjonsnavn: 'Advokatkontoret Aremark',
                    organisasjonsnummer: '01234567891011'
                },
                organisasjonSomAdressat: null,
                personSomAdressat: null
            },
            adresse: {
                linje1: 'Elgelia 20, 0000 Aremark',
                linje2: null,
                linje3: null,
                sistEndret: null,
                gyldighetsPeriode: null
            },
            registrert: '2010-02-02' as LocalDate,
            skifteform: Skifteform.OFFENTLIG,
            sistEndret: {
                ident: 'Folkeregisteret',
                tidspunkt: '2020-01-01T10:15:30' as LocalDateTime,
                system: 'Folkeregisteret'
            }
        }
    ],
    fullmakt: [
        {
            motpartsPersonident: '123456789',
            motpartsPersonNavn: {
                fornavn: 'Navn',
                mellomnavn: null,
                etternavn: 'Navnesen'
            },
            motpartsRolle: FullmaktsRolle.FULLMEKTIG,
            omrade: [
                {
                    kode: '*',
                    beskrivelse: ''
                }
            ],
            gyldighetsPeriode: {
                gyldigFraOgMed: '2021-06-01' as LocalDate,
                gyldigTilOgMed: '2022-05-01' as LocalDate
            }
        }
    ],
    vergemal: [
        {
            ident: '21042900076',
            navn: {
                fornavn: 'Simen',
                mellomnavn: null,
                etternavn: 'Solli'
            },
            vergesakstype: 'Voksen',
            omfang: 'Ivareta personens interesser innenfor det personlige og økonomiske området',
            embete: 'Fylkesmannen i Troms og Finnmark',
            gyldighetsPeriode: {
                gyldigFraOgMed: '2021-06-01' as LocalDate,
                gyldigTilOgMed: '2022-05-10' as LocalDate
            }
        },
        {
            ident: '123456799',
            navn: {
                fornavn: 'Truls',
                mellomnavn: null,
                etternavn: 'Tøffel'
            },
            vergesakstype: 'Fremtidsfullmakt',
            omfang: 'Ivareta personens interesser innenfor det økonomiske området',
            embete: 'Fylkesmannen i Troms og Finnmark',
            gyldighetsPeriode: {
                gyldigFraOgMed: '2021-02-10' as LocalDate,
                gyldigTilOgMed: '2022-05-01' as LocalDate
            }
        }
    ],
    tilrettelagtKommunikasjon: {
        talesprak: [
            {
                kode: 'NO',
                beskrivelse: 'Norsk'
            }
        ],
        tegnsprak: [
            {
                kode: 'NO',
                beskrivelse: 'Norsk'
            }
        ]
    },
    telefonnummer: [
        {
            retningsnummer: {
                kode: '+47',
                beskrivelse: 'Norge'
            },
            identifikator: '99009900',
            sistEndret: {
                ident: '11223344',
                tidspunkt: '2018-06-14T00:00:00' as LocalDateTime,
                system: 'BD06'
            },
            prioritet: -1
        }
    ],
    kontaktOgReservasjon: {
        personident: '10108000398',
        reservasjon: 'true',
        epostadresse: {
            value: 'epost@nav.no',
            sistOppdatert: null,
            sistVerifisert: '2013-01-01' as LocalDate
        },
        mobiltelefonnummer: {
            value: '12345678',
            sistOppdatert: '2021-11-02' as LocalDate,
            sistVerifisert: '2021-11-02' as LocalDate
        }
    },
    bankkonto: {
        kontonummer: '12345678900',
        banknavn: 'DNB ASA',
        sistEndret: {
            ident: '1010800 BD03',
            tidspunkt: '2006-03-15T00:00:00' as LocalDateTime,
            system: ''
        },
        bankkode: null,
        swift: 'DNBANOKKXXX',
        landkode: null,
        adresse: {
            linje1: 'Bankveien 1,',
            linje2: '0357 Bankestad',
            linje3: null,
            sistEndret: null,
            gyldighetsPeriode: null
        },
        valuta: {
            kode: 'NOK',
            beskrivelse: 'Norske kroner'
        }
    },
    forelderBarnRelasjon: hentBarnAremark()
};
