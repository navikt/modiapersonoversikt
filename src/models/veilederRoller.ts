export interface VeilederRoller {
    roller: string[];
}
