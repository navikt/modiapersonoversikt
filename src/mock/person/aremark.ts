import { Diskresjonskoder, TilrettelagtKommunikasjonsTyper } from '../../konstanter';
import { BostatusTyper, Kjønn, Person, Relasjonstype, SivilstandTyper } from '../../models/person/person';
import { SikkerhetstiltakTyper } from '../../models/sikkerhetstiltak';
import { Personadresse } from '../../models/personadresse';

export const gateAdresseAremark: Personadresse = {
    endringsinfo: {
        sistEndret: '2017-11-25T21:29:14.049Z',
        sistEndretAv: 'SKD'
    },
    gateadresse: {
        gatenavn: 'Islandsgate',
        husnummer: '49',
        postnummer: '7693',
        poststed: 'Hovedøya'
    }
};

export const postboksAdresseAremark: Personadresse = {
    endringsinfo: {
        sistEndret: '2018-04-10T09:06:58.256Z',
        sistEndretAv: 'SKD'
    },
    postboksadresse: {
        postboksnummer: '9698',
        postnummer: '7705',
        poststed: 'Færder fyr'
    }
};

export const matrikkelAdresseAremark: Personadresse = {
    endringsinfo: {
        sistEndret: '2018-04-10T09:06:58.256Z',
        sistEndretAv: 'SKD'
    },
    matrikkeladresse: {
        postnummer: '5468',
        poststed: 'Lillomarka'
    }
};

export const utlandsAdresseAremark: Personadresse = {
    endringsinfo: {
        sistEndret: '2018-04-10T09:06:58.256Z',
        sistEndretAv: 'SKD'
    },
    utlandsadresse: {
        landkode: {
            kodeRef: 'SYD',
            beskrivelse: 'Syden'
        },
        adresselinjer: ['Ut av landet', 'Rundt svingen']
    }
};

export const aremark: Person = {
    fødselsnummer: '10108000398',
    kjønn: Kjønn.Mann,
    geografiskTilknytning: '0118',
    alder: 42,
    folkeregistrertAdresse: gateAdresseAremark,
    alternativAdresse: postboksAdresseAremark,
    bankkonto: {
        banknavn: 'Bien sparebank',
        kontonummer: '12345678911',
        sistEndret: '2005-12-12',
        sistEndretAv: 'z9910',
        landkode: {
            kodeRef: 'NOR',
            beskrivelse: 'Norge'
        },
        valuta: {
            beskrivelse: 'Spesidaler',
            kodeRef: 'SPD'
        }
    },
    sikkerhetstiltak: {
        sikkerhetstiltaksbeskrivelse:
            'Får verken møte opp på NAV-kontor eller ringe på telefon ' + 'grunnet truende oppførsel og språk.',
        sikkerhetstiltakskode: SikkerhetstiltakTyper.FysiskOgTelefonUtestengelse,
        periode: {
            fra: '2005-12-12',
            til: '2025-12-12'
        }
    },
    navn: {
        endringsinfo: {
            sistEndret: '2012-12-12',
            sistEndretAv: 'NAV'
        },
        sammensatt: 'AREMARK TESTFAMILIEN',
        fornavn: 'AREMARK',
        mellomnavn: '',
        etternavn: 'TESTFAMILIEN'
    },
    tilrettelagtKomunikasjonsListe: [
        {
            kodeRef: 'TOHJ',
            beskrivelse: TilrettelagtKommunikasjonsTyper.TOHJ
        }
    ],
    diskresjonskode: { kodeRef: Diskresjonskoder.FORTROLIG_ADRESSE, beskrivelse: 'Sperret adresse, fortrolig' },
    statsborgerskap: { kodeRef: 'NOR', beskrivelse: 'NORGE' },
    personstatus: {
        dødsdato: undefined,
        bostatus: {
            kodeRef: BostatusTyper.Utvandret,
            beskrivelse: BostatusTyper.Utvandret
        }
    },
    sivilstand: {
        kodeRef: SivilstandTyper.Gift,
        beskrivelse: 'Gift',
        fraOgMed: '2005-12-12'
    },
    familierelasjoner: [
        {
            rolle: Relasjonstype.Barn,
            tilPerson: {
                navn: null,
                diskresjonskode: {
                    kodeRef: Diskresjonskoder.FORTROLIG_ADRESSE,
                    beskrivelse: 'Sperret adresse, fortrolig'
                },
                personstatus: {}
            }
        },
        {
            rolle: Relasjonstype.Barn,
            tilPerson: {
                navn: null,
                diskresjonskode: {
                    kodeRef: Diskresjonskoder.STRENGT_FORTROLIG_ADRESSE,
                    beskrivelse: 'Sperret adresse, strengt fortrolig'
                },
                personstatus: {}
            }
        },
        {
            harSammeBosted: true,
            rolle: Relasjonstype.Barn,
            tilPerson: {
                fødselsnummer: '10117100000',
                navn: {
                    sammensatt: 'Siri Bellomnavn Aremark',
                    fornavn: 'Siri',
                    mellomnavn: 'Bellomnavn',
                    etternavn: 'Aremark'
                },
                personstatus: {},
                alder: 15
            }
        },
        {
            harSammeBosted: false,
            rolle: Relasjonstype.Barn,
            tilPerson: {
                fødselsnummer: '10117100100',
                navn: {
                    sammensatt: 'Ola Ballomnavn Aremark',
                    fornavn: 'Ola',
                    mellomnavn: 'Ballomnavn',
                    etternavn: 'Aremark'
                },
                personstatus: {},
                alderMåneder: 5
            }
        },
        {
            harSammeBosted: false,
            rolle: Relasjonstype.Ektefelle,
            tilPerson: {
                navn: null,
                diskresjonskode: {
                    kodeRef: Diskresjonskoder.STRENGT_FORTROLIG_ADRESSE,
                    beskrivelse: 'Sperret adresse, strengt fortrolig'
                },
                personstatus: {}
            }
        }
    ],
    kontaktinformasjon: {
        mobil: {
            sistEndret: '2014-06-21T18:44:39+02:00',
            identifikator: '99887766',
            retningsnummer: { kodeRef: '+47', beskrivelse: 'Norge' },
            sistEndretAv: '10108000398'
        }
    }
};
