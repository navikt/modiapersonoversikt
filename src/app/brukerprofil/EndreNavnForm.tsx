import * as React from 'react';
import { ChangeEvent, FormEvent } from 'react';
import { Action } from 'history';
import { connect, Dispatch } from 'react-redux';

import Input from 'nav-frontend-skjema/lib/input';
import Undertittel from 'nav-frontend-typografi/lib/undertittel';
import KnappBase from 'nav-frontend-knapper';

import { STATUS } from '../../redux/utils';
import { EndreNavnRequest } from '../../redux/brukerprofil/endreNavnRequest';
import { Person } from '../../models/person/person';
import { AppState } from '../../redux/reducer';
import { endreNavn, reset } from '../../redux/brukerprofil/endreNavn';
import { VeilederRoller } from '../../models/veilederRoller';
import RequestTilbakemelding from './kontaktinformasjon/RequestTilbakemelding';

interface State {
    fornavnInput: string;
    mellomnavnInput: string;
    etternavnInput: string;
}

interface DispatchProps {
    endreNavn: (request: EndreNavnRequest) => void;
    resetEndreNavnReducer: () => void;
}

interface StateProps {
    status: STATUS;
}

interface OwnProps {
    person: Person;
    veilederRoller?: VeilederRoller;
}

type Props = DispatchProps & StateProps & OwnProps;

class EndreNavnForm extends React.Component<Props, State> {

    constructor(props: Props) {
        super(props);

        this.state = {
            fornavnInput: props.person.navn.fornavn || '',
            mellomnavnInput: props.person.navn.mellomnavn || '',
            etternavnInput: props.person.navn.etternavn || ''
        };

        this.fornavnInputChange = this.fornavnInputChange.bind(this);
        this.mellomnavnInputChange = this.mellomnavnInputChange.bind(this);
        this.etternavnInputChange = this.etternavnInputChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    componentWillUnmount() {
        this.props.resetEndreNavnReducer();
    }

    fornavnInputChange(event: ChangeEvent<HTMLInputElement>) {
        this.setState({
            fornavnInput: event.target.value
        });
        this.props.resetEndreNavnReducer();
    }

    mellomnavnInputChange(event: ChangeEvent<HTMLInputElement>) {
        this.setState({
            mellomnavnInput: event.target.value
        });
        this.props.resetEndreNavnReducer();

    }

    etternavnInputChange(event: ChangeEvent<HTMLInputElement>) {
        this.setState({
            etternavnInput: event.target.value
        });
        this.props.resetEndreNavnReducer();
    }

    navnErEndret() {
        const fornavnErEndret = this.state.fornavnInput !== this.props.person.navn.fornavn;
        const mellomnavnErEndret = this.state.mellomnavnInput !== this.props.person.navn.mellomnavn;
        const etternavnErEndret = this.state.etternavnInput !== this.props.person.navn.etternavn;
        return fornavnErEndret || mellomnavnErEndret || etternavnErEndret;
    }

    handleSubmit(event: FormEvent<HTMLFormElement>) {
        this.props.endreNavn({
            fødselsnummer: this.props.person.fødselsnummer,
            fornavn: this.state.fornavnInput,
            mellomnavn: this.state.mellomnavnInput,
            etternavn: this.state.etternavnInput
        });
        event.preventDefault();
    }

    harVeilderPåkrevdRolle() {
        if (!this.props.veilederRoller) {
            return false;
        }
        return this.props.veilederRoller.roller.includes('0000-GA-BD06_EndreNavn');
    }

    render() {
        const harIkkeTilgang = !this.harVeilderPåkrevdRolle();
        return (
            <form onSubmit={this.handleSubmit}>
                <Undertittel>Navn</Undertittel>
                <Input
                    label="Fornavn"
                    value={this.state.fornavnInput}
                    onChange={this.fornavnInputChange}
                    disabled={harIkkeTilgang}
                />
                <Input
                    label="Mellomnavn"
                    value={this.state.mellomnavnInput}
                    onChange={this.mellomnavnInputChange}
                    disabled={harIkkeTilgang}
                />
                <Input
                    label="Etternavn"
                    value={this.state.etternavnInput}
                    onChange={this.etternavnInputChange}
                    disabled={harIkkeTilgang}
                />
                <KnappBase
                    type="standard"
                    spinner={this.props.status === STATUS.PENDING}
                    disabled={harIkkeTilgang || !this.navnErEndret()}
                    autoDisableVedSpinner={true}
                >
                    Endre navn
                </KnappBase>
                <RequestTilbakemelding
                    status={this.props.status}
                    onSuccess={'Navnet ble endret. Det kan ta noen minutter før endringene blir synlig.'}
                    onError={'Det skjedde en feil ved endring av navn.'}
                />
            </form>

        );
    }
}

const mapStateToProps = (state: AppState): StateProps => {
    return ({
        status: state.endreNavn.status
    });
};

function mapDispatchToProps(dispatch: Dispatch<Action>): DispatchProps {
    return {
        endreNavn: (request: EndreNavnRequest) => dispatch(endreNavn(request)),
        resetEndreNavnReducer: () => dispatch(reset())
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(EndreNavnForm);
