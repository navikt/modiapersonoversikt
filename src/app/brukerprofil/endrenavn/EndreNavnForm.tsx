import * as React from 'react';
import { FormEvent } from 'react';
import { connect } from 'react-redux';

import { Input } from 'nav-frontend-skjema';
import { Undertittel } from 'nav-frontend-typografi';
import KnappBase from 'nav-frontend-knapper';
import { EndreNavnRequest } from '../../../redux/restReducers/brukerprofil/endreNavnRequest';
import { Person } from '../../../models/person/person';
import { AppState } from '../../../redux/reducers';
import { VeilederRoller } from '../../../models/veilederRoller';
import { FormKnapperWrapper } from '../BrukerprofilForm';
import RequestTilbakemelding from '../RequestTilbakemelding';
import { brukersNavnKanEndres, validerNavn } from './endrenavn-utils';
import { EndreNavnInfomeldingWrapper } from '../Infomelding';
import { ignoreEnter, InputState, visEndringsinfo } from '../utils/formUtils';
import { FormFieldSet } from '../../personside/visittkort/body/VisittkortStyles';
import { veilederHarPåkrevdRolleForEndreNavn } from '../../../utils/RollerUtils';
import { loggEvent } from '../../../utils/frontendLogger';
import { AsyncDispatch } from '../../../redux/ThunkTypes';
import { PostStatus } from '../../../rest/utils/postResource';

interface NavnInputProps {
    label: string;
    state: InputState;
    onChange: (input: string) => void;
}

function NavnInput({ label, state, onChange }: NavnInputProps) {
    const feilmelding = state.feilmelding ? { feilmelding: state.feilmelding } : undefined;
    return (
        <Input
            label={label}
            name={label}
            value={state.input}
            onChange={event => onChange(event.target.value.toUpperCase())}
            onKeyPress={ignoreEnter}
            feil={feilmelding}
        />
    );
}

interface State {
    fornavn: InputState;
    mellomnavn: InputState;
    etternavn: InputState;
    formErEndret: boolean;
}

interface DispatchProps {
    endreNavn: (request: EndreNavnRequest) => void;
    resetEndreNavnResource: () => void;
}

interface StateProps {
    resourceStatus: PostStatus;
}

interface OwnProps {
    person: Person;
    veilederRoller: VeilederRoller;
}

type Props = DispatchProps & StateProps & OwnProps;

class EndreNavnForm extends React.Component<Props, State> {
    constructor(props: Props) {
        super(props);

        this.state = this.initialState(props);

        this.fornavnInputChange = this.fornavnInputChange.bind(this);
        this.mellomnavnInputChange = this.mellomnavnInputChange.bind(this);
        this.etternavnInputChange = this.etternavnInputChange.bind(this);
        this.tilbakestillForm = this.tilbakestillForm.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    initialState(props: Props) {
        return {
            fornavn: {
                input: props.person.navn.fornavn || '',
                feilmelding: null
            },
            mellomnavn: {
                input: props.person.navn.mellomnavn || '',
                feilmelding: null
            },
            etternavn: {
                input: props.person.navn.etternavn || '',
                feilmelding: null
            },
            formErEndret: false
        };
    }

    componentWillUnmount() {
        this.props.resetEndreNavnResource();
    }

    fornavnInputChange(input: string) {
        this.setState({
            fornavn: {
                input,
                feilmelding: null
            },
            formErEndret: true
        });
    }

    mellomnavnInputChange(input: string) {
        this.setState({
            mellomnavn: {
                feilmelding: null,
                input
            },
            formErEndret: true
        });
    }

    etternavnInputChange(input: string) {
        this.setState({
            etternavn: {
                feilmelding: null,
                input
            },
            formErEndret: true
        });
    }

    navnErEndret() {
        const fornavnErEndret = this.state.fornavn.input !== this.props.person.navn.fornavn;
        const mellomnavnErEndret = this.state.mellomnavn.input !== this.props.person.navn.mellomnavn;
        const etternavnErEndret = this.state.etternavn.input !== this.props.person.navn.etternavn;
        return fornavnErEndret || mellomnavnErEndret || etternavnErEndret;
    }

    tilbakestillForm(event: React.MouseEvent<HTMLButtonElement>) {
        this.setState(this.initialState(this.props));
        this.props.resetEndreNavnResource();
        event.preventDefault();
    }

    validerFornavn() {
        try {
            validerNavn(this.state.fornavn.input);
        } catch (error) {
            this.setState({
                fornavn: {
                    ...this.state.fornavn,
                    feilmelding: error.message
                }
            });
            throw error;
        }
    }

    validerEtternavn() {
        try {
            validerNavn(this.state.etternavn.input);
        } catch (error) {
            this.setState({
                etternavn: {
                    ...this.state.etternavn,
                    feilmelding: error.message
                }
            });
            throw error;
        }
    }

    validerInput() {
        let validInput = true;

        try {
            this.validerFornavn();
        } catch {
            validInput = false;
        }

        try {
            this.validerEtternavn();
        } catch {
            validInput = false;
        }

        return validInput;
    }

    handleSubmit(event: FormEvent<HTMLFormElement>) {
        event.preventDefault();

        if (!this.validerInput()) {
            return;
        }

        this.setState({
            formErEndret: false
        });

        this.props.endreNavn({
            fødselsnummer: this.props.person.fødselsnummer,
            fornavn: this.state.fornavn.input,
            mellomnavn: this.state.mellomnavn.input,
            etternavn: this.state.etternavn.input
        });
        loggEvent('Endre navn', 'Brukerprofil');
    }

    render() {
        const kanEndreNavn =
            veilederHarPåkrevdRolleForEndreNavn(this.props.veilederRoller) && brukersNavnKanEndres(this.props.person);
        return (
            <form onSubmit={this.handleSubmit}>
                <FormFieldSet disabled={!kanEndreNavn}>
                    <Undertittel>Navn</Undertittel>
                    {visEndringsinfo(this.props.person.navn.endringsinfo)}
                    <EndreNavnInfomeldingWrapper person={this.props.person} veilderRoller={this.props.veilederRoller} />
                    <NavnInput label="Fornavn" state={this.state.fornavn} onChange={this.fornavnInputChange} />
                    <NavnInput label="Mellomnavn" state={this.state.mellomnavn} onChange={this.mellomnavnInputChange} />
                    <NavnInput label="Etternavn" state={this.state.etternavn} onChange={this.etternavnInputChange} />
                    <FormKnapperWrapper>
                        <KnappBase
                            type="standard"
                            onClick={this.tilbakestillForm}
                            disabled={!kanEndreNavn || !this.state.formErEndret || !this.navnErEndret()}
                        >
                            Avbryt
                        </KnappBase>
                        <KnappBase
                            type="hoved"
                            spinner={this.props.resourceStatus === PostStatus.POSTING}
                            disabled={!kanEndreNavn || !this.state.formErEndret || !this.navnErEndret()}
                            autoDisableVedSpinner={true}
                        >
                            Endre navn
                        </KnappBase>
                    </FormKnapperWrapper>
                    {!this.state.formErEndret ? (
                        <RequestTilbakemelding
                            status={this.props.resourceStatus}
                            onSuccess={'Navnet ble endret. Det kan ta noe tid før endringen blir synlig'}
                            onError={'Det skjedde en feil ved endring av navn.'}
                        />
                    ) : null}
                </FormFieldSet>
            </form>
        );
    }
}

const mapStateToProps = (state: AppState): StateProps => {
    return {
        resourceStatus: state.restResources.endreNavn.status
    };
};

function mapDispatchToProps(dispatch: AsyncDispatch): DispatchProps {
    return {
        endreNavn: (request: EndreNavnRequest) =>
            dispatch((d, getState) => d(getState().restResources.endreNavn.actions.post(request))),
        resetEndreNavnResource: () => dispatch((d, getState) => d(getState().restResources.endreNavn.actions.reset))
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(EndreNavnForm);
