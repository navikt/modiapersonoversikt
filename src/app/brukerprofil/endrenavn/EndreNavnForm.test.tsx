import * as React from 'react';
import * as renderer from 'react-test-renderer';
import TestProvider from '../../../test/Testprovider';
import { aremark } from '../../../mock/person/aremark';
import EndreNavnForm from './EndreNavnForm';
import { VeilederRoller } from '../../../models/veilederRoller';

test('Viser form for endre navn', () => {
    const veilederRoller: VeilederRoller = { roller: ['0000-GA-BD06_EndreNavn'] };
    const endreNavn = renderer.create(
        <TestProvider>
            <EndreNavnForm person={aremark} veilederRoller={veilederRoller} />
        </TestProvider>
    );

    expect(endreNavn.toJSON()).toMatchSnapshot();
});
