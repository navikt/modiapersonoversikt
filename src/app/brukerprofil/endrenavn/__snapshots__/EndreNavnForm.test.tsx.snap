// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Viser form for endre navn 1`] = `
.c1 {
  color: #78706a;
  margin-top: 3px;
}

.c1 .typo-etikett-liten {
  line-height: 1rem;
}

.c0 {
  margin: 0;
  padding: 0;
  border: 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c2 > * {
  margin: 0.5rem 0 0.5rem 1em;
}

<form
  onSubmit={[Function]}
>
  <fieldset
    className="c0"
    disabled={false}
  >
    <h2
      className="typo-undertittel"
    >
      Navn
    </h2>
    <div
      className="c1"
    >
      <p
        className="typo-etikett-liten"
      >
        Endret 
        12.12.2012
         
        NAV
      </p>
    </div>
    <div
      className="skjemaelement"
    >
      <label
        className="skjemaelement__label"
        htmlFor="Fornavn"
      >
        Fornavn
      </label>
      <input
        className="skjemaelement__input input--fullbredde"
        id="Fornavn"
        name="Fornavn"
        onChange={[Function]}
        onKeyPress={[Function]}
        type="text"
        value="AREMARK"
      />
      <div
        aria-live="assertive"
        role="alert"
      />
    </div>
    <div
      className="skjemaelement"
    >
      <label
        className="skjemaelement__label"
        htmlFor="Mellomnavn"
      >
        Mellomnavn
      </label>
      <input
        className="skjemaelement__input input--fullbredde"
        id="Mellomnavn"
        name="Mellomnavn"
        onChange={[Function]}
        onKeyPress={[Function]}
        type="text"
        value=""
      />
      <div
        aria-live="assertive"
        role="alert"
      />
    </div>
    <div
      className="skjemaelement"
    >
      <label
        className="skjemaelement__label"
        htmlFor="Etternavn"
      >
        Etternavn
      </label>
      <input
        className="skjemaelement__input input--fullbredde"
        id="Etternavn"
        name="Etternavn"
        onChange={[Function]}
        onKeyPress={[Function]}
        type="text"
        value="TESTFAMILIEN"
      />
      <div
        aria-live="assertive"
        role="alert"
      />
    </div>
    <div
      className="c2"
    >
      <button
        className="knapp knapp--standard knapp--disabled"
        disabled={true}
        onClick={[Function]}
        type="submit"
      >
        Avbryt
      </button>
      <button
        className="knapp knapp--hoved knapp--disabled"
        disabled={true}
        type="submit"
      >
        Endre navn
      </button>
    </div>
  </fieldset>
</form>
`;
